server/WFCORE-4251_list_modules_in_deployment_classpath.adoc

```diff
@@ -0,0 +1,118 @@
+* This information could facilitate discovering unneeded subsystems in the installation of his servers.
+
+
+The proposed solution is to include a new operation `list-modules` available under `/deployment` and, in case of ears, `/deployment=*/subdeployment` resource:
+
+ [standalone@localhost:9990 /] /deployment=ejb-in-ear.ear:list-modules
+
+ [standalone@localhost:9990 /] /deployment=ejb-in-ear.ear/subdeployment=ejb-in-ear-web.war:list-modules
+
+An example of the information returned by this method could be:
+
+  {
+      "outcome" => "success",
+      "result" => {
+          "ejb-in-ear.ear" => [
+              {
+                  "name" => "com.fasterxml.jackson.datatype.jackson-datatype-jdk8",
+                  "slot" => "main",
+                  "optional" => true,
+                  "export" => false,
+                  "import-services" => true,
+                  "duplicated" => false
+              },
+              {
+                  "name" => "com.fasterxml.jackson.datatype.jackson-datatype-jsr310",
+                  "slot" => "main",
+                  "optional" => true,
+                  "export" => false,
+                  "import-services" => true,
+                  "duplicated" => true
+              },
+              ...
+          ]
+      }
+  }
+
+Where:
+
+* name: The module name.
+* slot: The slot where this module is.
+* optional: If the dependency was added as an optional dependency.
+* export: If the dependency is being exported to other modules.
+* import-services: If this module is importing services from other dependencies.
+* duplicated: If it is a module defined by the user and the server is adding a module implicitly with the same name.
+
+
+== Issue Metadata
+
+=== Issue
+
+* {issue-base-url}WFCORE-4251[WFCORE-4251]
+
+=== Related Issues
+
+* {issue-base-url}EAP7-521[EAP7-521]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:szhantem@redhat.com[Sultan Zhantemirov]
+
+=== Affected Projects or Components
+
+* Wildfly Core
+* JBoss Modules (optional, it could help if we finally show the artifacts included on each module, see Nice-to-Have requirements)
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* It must be possible to list the implicit modules added to the classpath for a specific deployment and, in case of ears, for a particular sub-deployment.
+* List of modules should be sorted alphabetically by module name.
+* The list of modules will be only available if the deployment is enabled.
+* The new operation will be available only in standalone mode. Domain mode changes are not going to be covered by this RFE.
+* The output of the operation will include an attribute named `duplicated` which will be true if the module was added by the user in the deployment, via manifest file or deployment-structure.xml, and added implicitly by the server.
```

Comment:
If we have the data to know if it was duplicate, how about an implicit=true|false?  Probably some better name than implicit, more descriptive.

TBH, 'duplicated' sounds like a nice-to-have. I doubt that it's a common occurrence and AFAIK it's harmless if it occurs. This check would not cover cases where the deployment includes a jar that is also available via a module dependency. I expect that is much more common and is more likely to be a problem, as it means classes with the same name are available from two different classloaders. (I don't think this RFE should try and address that though!)

Reply:
> If we have the data to know if it was duplicate, how about an implicit=true|false? Probably some better name than implicit, more descriptive.

Do you mean to add additionally a new attribute implicit=true|false so if 'duplicate=false', then to know if it was a dependency added by the server or by the user?, if yes, we can also use a user-defined=true|false

> 
> Separate point:
> 
> TBH, 'duplicated' sounds like a nice-to-have. I doubt that it's a common occurrence and AFAIK it's harmless if it occurs. 

Agreed.

> This check would not cover cases where the deployment includes a jar that is also available via a module dependency. I expect that is much more common and is more likely to be a problem, as it means classes with the same name are available from two different classloaders. (I don't think this RFE should try and address that though!)

Yes, the idea is to be able to know if the user is including modules explicitly via manifest file or deployment-structure.xml when they are not required because the server is including them implicitly.

Reply:
I like 'user-defined'. 👍

Reply:
If we are displaying the information divided into categories like system dependencies, user dependencies, and local dependencies, the user-defined attribute is no longer needed.

Reply:
Be cautious about exposing implementation details.

Reply:
@dmlloyd confirmed that the intention is to have a long term support for the following attributes exposed by this operation:
- name
- slot
- optional
- export
- import-services

Additionally he suggested to collapse "name" and "slot" in a single "name" property. A module with a slot of "main" would just be named by the module name; other slots would be concatenated to the end of the module name after the ":" separator.

I'm going to update the proposal with this last point.